### PR TITLE
Changes the shrike's abilities

### PR DESCRIPTION
**Pull request by Spyroshark**
## About The Pull Request

This PR changes how the shrike's fling works, and replaces Psychic Choke with a new ability, "Unrelenting Force". It also adds a "Shake" proc to make an atom shake around, which is used for the new ability.

Psychic Fling changes:
Psychic fling now only works two tiles away from your target, but can also fling items, and items or people you use it on will be sent further away.

Unrelenting Force:
When this ability is used, everything in a 2x3 radius in front of the shrike will be violently flung back. This works on both people and items, and applies a short stun like the fling currently does. It also applies a "shaking" animation to the ground and other terrain objects.

## Why It's Good For The Game

Right now both of the shrike's offensive abilities are ranged stuns, and are disliked by most players. The new ability will (hopefully) not be as powerful, and because both of the shrike's abilities won't work past 2 tiles, it should reduce the frustration of being stunned from offscreen. Unrelenting Force might need some rebalancing if it feels too powerful, though.

## Changelog
:cl:
add: The shrike has a new ability, "Unrelenting Force" that sends everything in front of the shrike flying away.
tweak: Psychic Fling now only works from 2 tiles away, but can fling items as well as humans.
tweak: Psychic Choke has been replaced with Unrelenting Force.
/:cl:

### CODE REVIEW

File: code/datums/keybinding/xeno.dm [hunk 1]
```diff
@@ -335,6 +335,13 @@
 	description = ""
 	keybind_signal = COMSIG_XENOABILITY_PSYCHIC_FLING
 
+/datum/keybinding/xeno/unrelenting_force
+	key = "Unbound"
+	name = "unrelenting_force"
+	full_name = "Unrelenting Force"
+	description = ""
+	keybind_signal = COMSIG_XENOABILITY_PSYCHIC_FLING
```
Comment (Rohesie) [inline code/datums/keybinding/xeno.dm]: It should have its own signal.

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 2]
```diff
@@ -73,17 +73,17 @@
 		return FALSE
 	if(QDELETED(target))
 		return FALSE
-	if(!ishuman(target))
+	if(!istype(target, /obj/item) && !ishuman(target))	//only items and mobs can be flung.
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	if(!isitem(target) && !ishuman(target))	//only items and mobs can be flung.
```

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 3]
```diff
@@ -92,18 +92,19 @@
...
 	add_cooldown()
-	victim.apply_effects(1, 2) 	// Stun
-	shake_camera(victim, 2, 1)
+	if(ishuman(victim))
+		victim.apply_effects(1, 2) 	// Stun
+		shake_camera(victim, 2, 1)
 
 	var/facing = get_dir(owner, victim)
-	var/fling_distance = 3
+	var/fling_distance = (!ishuman(victim)) ? 4 : 3 //Objects get flung further away.
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	var/fling_distance = (isitem(victim)) ? 4 : 3 //Objects get flung further away.
```
A bit more clear.

File: code/__HELPERS/unsorted.dm [hunk 4]
```diff
@@ -1143,6 +1143,14 @@ GLOBAL_LIST_INIT(wallitems, typecacheof(list(
 			assembled += A
 	return assembled
 
+/atom/proc/Shake(pixelshiftx = 15, pixelshifty = 15, duration = 250) //Does a "shaking" effect on a sprite, code is from tgstation
```
Comment (Rohesie) [inline code/__HELPERS/unsorted.dm]:
```suggestion
/atom/proc/Shake(pixelshiftx = 15, pixelshifty = 15, duration = 25 SECONDS) //Does a "shaking" effect on a sprite, code is from tgstation
```
This is measured in deciseconds, no?

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 5]
```diff
@@ -115,6 +116,65 @@
...
+			upper_right = locate(owner.x + 1, owner.y - 1, owner.z)
+		if(WEST)
+			lower_left = locate(owner.x - 2, owner.y - 1, owner.z)
+			upper_right = locate(owner.x - 1, owner.y + 1, owner.z)
+		if(EAST)
+			lower_left = locate(owner.x + 1, owner.y - 1, owner.z)
+			upper_right = locate(owner.x + 2, owner.y + 1, owner.z)
+
+	for(var/turf/affected_tile in block(lower_left, upper_right)) //everything in the 2x3 block is found.
+		affected_tile.Shake(4, 4, 20)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
		affected_tile.Shake(4, 4, 2 SECONDS)
```
Is this the proper time?

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 6]
```diff
@@ -115,6 +116,65 @@
...
+		for(var/atom/movable/target in affected_tile)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
		for(var/i in affected_tile)
		var/atom/movable/target = i
```
There can only be movable atoms in a tile, so skip the check.

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 7]
```diff
@@ -115,6 +116,65 @@
...
+			target.throw_at(throwlocation, 6, 1, owner, TRUE)
+			if(ishuman(target)) //if they're human, they also should get knocked off their feet from the blast.
+				var/mob/living/carbon/human = target
+				human.apply_effects(1, 2) 	// Stun
+				shake_camera(target, 2, 1)
+
+	owner.visible_message("<span class='xenowarning'>[owner] sends out a huge blast of psychic energy!</span>", \
+	"<span class='xenowarning'>We send out a huge blast of psychic energy!</span>")
+
+	playsound(owner,'sound/effects/alien_tail_swipe3.ogg', 75, 1)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	playsound(owner,'sound/effects/alien_tail_swipe3.ogg', 75, TRUE)
```

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm [hunk 8]
```diff
@@ -65,3 +65,12 @@
...
 		new_hud.remove_hud_from(usr)
+
+
+// ***************************************
+// *********** Icon
+// ***************************************
+/mob/living/carbon/xenomorph/shrike/proc/stop_screeching()
+	if(icon_state != "Shrike Screeching")
+		return
+	icon_state = "Shrike Running"
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm]:
```suggestion
	icon_state = "Shrike Running"

```
Try ending files with a new line.

File: code/datums/keybinding/xeno.dm (hunk 1 above)
Comment (Spyroshark) [reply code/datums/keybinding/xeno.dm]: i was pretty tired when i was adding the keybindings, must of missed that

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm (hunk 2 above)
Comment (Spyroshark) [reply code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]: ah, wasn't aware we had that

File: code/__HELPERS/unsorted.dm (hunk 4 above)
Comment (Spyroshark) [reply code/__HELPERS/unsorted.dm]: I wasn't sure if it should be measured in seconds or not, i'll change it

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm (hunk 8 above)
Comment (Spyroshark) [reply code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm]: oh, i didn't know i was supposed to do that

Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm]:
```suggestion
	update_icon()
```

**Comment by Spyroshark** [comment]
Small issue with the new ability, it does not make you turn to where you click. I'm going to fix it today, please don't merge/deny this yet.

**Comment by Spyroshark** [comment]
alright, now that ive seen it ingame a bit i made a couple changes.
Fling now has a 3 tile range, like non-ruler shrike used to.
Unrelenting force now turns you to where you clicked, and it also has a 3x3 radius instead of the 2x3. This should help with it being too hard to use.

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 9]
```diff
@@ -115,6 +116,68 @@
...
+
+
+/datum/action/xeno_action/activable/unrelenting_force/on_cooldown_finish()
+	to_chat(owner, "<span class='notice'>Our mind is ready to unleash another blast of force.</span>")
+	return ..()
+
+
+/datum/action/xeno_action/activable/unrelenting_force/use_ability(atom/target)
+	add_cooldown()
+	addtimer(CALLBACK(owner, /mob/living/carbon/xenomorph/shrike/proc/stop_screeching), 1 SECONDS)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	addtimer(CALLBACK(owner, /atom/movable.proc/stop_screeching), 1 SECONDS)
```

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm [hunk 10]
```diff
@@ -65,3 +65,12 @@
...
 		new_hud.remove_hud_from(usr)
+
+
+// ***************************************
+// *********** Icon
+// ***************************************
+/mob/living/carbon/xenomorph/shrike/proc/stop_screeching()
+	if(icon_state != "Shrike Screeching")
+		return
+	update_icon()
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/shrike.dm]: Delete this whole thing

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 11]
```diff
@@ -115,6 +116,68 @@
...
+
+
+/datum/action/xeno_action/activable/unrelenting_force/on_cooldown_finish()
+	to_chat(owner, "<span class='notice'>Our mind is ready to unleash another blast of force.</span>")
+	return ..()
+
+
+/datum/action/xeno_action/activable/unrelenting_force/use_ability(atom/target)
+	add_cooldown()
+	addtimer(CALLBACK(owner, /atom/movable.proc/stop_screeching), 1 SECONDS)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	addtimer(CALLBACK(owner, /atom/movable.proc/update_icon), 1 SECONDS)
```

File: code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm [hunk 12]
```diff
@@ -115,6 +116,68 @@
...
+
+
+/datum/action/xeno_action/activable/unrelenting_force/on_cooldown_finish()
+	to_chat(owner, "<span class='notice'>Our mind is ready to unleash another blast of force.</span>")
+	return ..()
+
+
+/datum/action/xeno_action/activable/unrelenting_force/use_ability(atom/target)
+	add_cooldown()
+	addtimer(CALLBACK(owner, /atom/movable.proc/update_icon), 1 SECONDS)
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/shrike/abilities_shrike.dm]:
```suggestion
	addtimer(CALLBACK(owner, /mob.proc/update_icons), 1 SECONDS)
```

**Comment by spookydonut** [comment]
@Rohesie has this been tested? if so then merge

**Comment by spookydonut** [comment]
spyro says they have changes they need to do

**Pull request merged**